crates/store/src/db/models/queries/nullifiers.rs

```diff
@@ -23,23 +23,49 @@ use crate::db::models::conv::{SqlTypeConvert, nullifier_prefix_to_raw_sql};
 use crate::db::models::utils::{get_nullifier_prefix, vec_raw_try_into};
 use crate::db::{NullifierInfo, schema};
 
-/// Returns nullifiers filtered by prefix and block creation height.
+/// Returns nullifiers filtered by prefix within a block number range.
 ///
-/// Each value of the `nullifier_prefixes` is only the `prefix_len` most significant bits
-/// of the nullifier of interest to the client. This hides the details of the specific
-/// nullifier being requested. Currently the only supported prefix length is 16 bits.
+/// Each value of the `nullifier_prefixes` is only the `prefix_len` most significant bits of the
+/// nullifier of interest to the client. This hides the details of the specific nullifier being
+/// requested. Currently the only supported prefix length is 16 bits.
 ///
 /// # Returns
 ///
-/// A vector of [`NullifierInfo`] with the nullifiers and the block height at which they were
+/// Range and pagination semantics:
+/// - Both `block_from` and `block_to` are inclusive bounds.
+/// - To keep responses ≤ ~2MB, an internal row limit is enforced. When the limit is hit and the
+///   result spans multiple blocks, the last block in the page is dropped entirely, and
+///   `last_block_included` is set to the block number immediately before that dropped block.
+///   Clients should resume with `block_from = last_block_included + 1`.
+/// - If all rows belong to a single block and hit the limit, this function returns an empty
+///   `nullifiers` page with `last_block_included = that_block - 1`. Intra-block pagination is not
+///   supported; callers may need to narrow the range.
+///
+/// A tuple `(nullifiers, last_block_included)` where:
+/// - `nullifiers` is a vector of [`NullifierInfo`] (each contains the nullifier and the block
+///   number at which it was created), ordered by block number ascending.
+/// - `last_block_included` is the last block number fully included in this response. If the
+///   internal row limit is reached (to cap response size), this may be less than `block_to`. In
+///   that case, the caller should re-issue the query with `block_from = last_block_included + 1` to
+///   continue.
 pub(crate) fn select_nullifiers_by_prefix(
     conn: &mut SqliteConnection,
     prefix_len: u8,
     nullifier_prefixes: &[u16],
-    block_num: BlockNumber,
-) -> Result<Vec<NullifierInfo>, DatabaseError> {
+    block_from: BlockNumber,
+    block_to: BlockNumber,
+) -> Result<(Vec<NullifierInfo>, BlockNumber), DatabaseError> {
+    pub const MAX_PAYLOAD_BYTES: usize = 2 * 1024 * 1024; // 2 MB
+    pub const NULLIFIER_BYTES: usize = 32; // digest size
+    pub const ROW_OVERHEAD_BYTES: usize = NULLIFIER_BYTES + size_of::<i64>(); // 40 bytes
```

Comment:
Why do we add `size_of::<i64>()` here? The block number is 32 bits - so, shouldn't this be `size_of::<u32>()`?

Reply:
Updated!